### PR TITLE
Support for complex nested objects

### PR DESCRIPTION
**Pull request by KAllan357**
# Description of change
Adding support for custom complex objects.

After looking at the output generated from a more complex report it has led us to the conclusion that using a format of json and then using the xsds for schema generation will be ideal. Using `format=json` makes syncing the data easy and the xsds includes sufficient type information for the JSON to allow the transformer to properly cast decimals.

Part of this change includes manipulating the URL to force it to have a `format=json` query parameter. This will allow the tap to work without needing our customers to update their config.

# Manual QA steps
 - Created a new report with complex types and tested syncing json format
 
# Risks
 - There is a potential for column splitting and I can find no guarantee that the JSON format is available for all reports. We might get through this only to find that our customers don't have JSON enabled for some reason.
 - It is unclear whether the current implementation streams the JSON or if it downloads it all at the start. It's also unclear whether worrying about streaming the report is pre-emptive optimization and would not cause issues even if it was downloaded wholesale.
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_workday_raas/client.py [hunk 1]
```diff
@@ -1,16 +1,30 @@
...
+    # Force the format query param to be set to format=json
+
+    # Split query params off
+    url_breakdown = report_url.split('?')
+
+    # Gather all params that are not format
+    if len(url_breakdown) == 1:
+        params = []
+    else:
+        params = [x for x in url_breakdown[1].split('&') if not x.startswith('format')]
```
Comment (dmosorast) [inline tap_workday_raas/client.py]: This should probably `format=` to get the right hand side match.

File: tap_workday_raas/client.py [hunk 2]
```diff
@@ -1,16 +1,30 @@
...
+    params.append('format=json')
+    param_string = '&'.join(params)
+
+    # Put the url back together
+    corrected_url = url_breakdown[0] + '?' + param_string
+
+    # Get the data
+    with requests.get(corrected_url, auth=(user, password), stream=True) as resp:
+        resp.raise_for_status()
+        report = resp.json() # TODO Check that this is streaming. I don't think it is
```
Comment (dmosorast) [inline tap_workday_raas/client.py]: Talked about this already, but this should use a JSON streaming parser if we're concerned about memory usage for large reports.

File: tap_workday_raas/discover.py [hunk 3]
```diff
@@ -8,39 +8,94 @@
 
 LOGGER = singer.get_logger()
 
-def _type_to_schema(elem_type):
+def _element_to_schema(element):
+    elem_type = element.attrib['type'].split(':')[1]
+    is_nullable = element.attrib.get('minOccurs') == '0'
+    # NB: The only actual value I've seen for maxOccurs is "unbounded". There may be other possible values
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: In this case, it should probably be defensive and fail if it sees a value other than unbounded, that way we can account for it if it comes up.

File: tap_workday_raas/discover.py [hunk 4]
```diff
@@ -8,39 +8,94 @@
...
+            'type': ['string'],
             'format': 'date-time'
         }
     elif elem_type == 'decimal':
         #TODO What else do we need for decimal types?
-        return {
-            'type': ['number', 'null']
+        schema = {
+            'type': ['number'],
+            'format': 'decimal'
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: Maybe not just yet if this tap uses the transformer, but `"type": "string", "format": "singer-decimal"` would be good.

File: tap_workday_raas/discover.py [hunk 5]
```diff
@@ -8,39 +8,94 @@
...
+
+    if is_list:
+        schema = {
+            'type': 'array',
+            'items': schema
+        }
+
+    return schema
+
+# TODO: Test this logic
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: Has this been to-did? Are there any cases here where we'd want to hard fail if we see something we don't know about?

File: tap_workday_raas/discover.py [hunk 6]
```diff
@@ -8,39 +8,94 @@
...
 
-                    schema_type = _type_to_schema(elem_type)
+        # When elem's type attribute is a type defined as its own complexType - a nested object
+        if elem_type in complex_type_mapping:
 
-                    schema['properties'][elem_name] = {
-                        **schema_type
-                    }
+            # TODO Check if the element is a list and type it correctly
+            is_list = elem.attrib.get("maxOccurs") == 'unbounded'
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: This again, this should probably be a function with a standard behavior and throw if something unexpected occurs.

**Comment by asaf-erlich** [comment]
Minor comment, since this is open source. You included a lot of good information in the PR description, but that's not seen in the commit message itself. Can you add a commit message which pulls that in so that if someone reads the git history they see that? I know people could always lookup the PR from the commit obviously...

**Review by dmosorast — CHANGES_REQUESTED** [review]
Seems solid overall, just a few safety-oriented comments, and call-outs of TODOs

File: tap_workday_raas/client.py [hunk 7]
```diff
@@ -1,16 +1,30 @@
...
-                    yield (event, elem)
-            except ElementTree.ParseError as e:
-                raise Exception("Report URL {} does not parse as XML. Please ensure the integration is configured with the correct URL".format(report_url))
+    # Force the format query param to be set to format=json
+
+    # Split query params off
+    url_breakdown = report_url.split('?')
+
+    # Gather all params that are not format
+    if len(url_breakdown) == 1:
```
Comment (asaf-erlich) [inline tap_workday_raas/client.py:11]: So if there is no question mark in the report url at all, would this if statement cause any issues? Should this condition be `<= 1`?

File: tap_workday_raas/client.py (hunk 2 above)
Comment (asaf-erlich) [inline tap_workday_raas/client.py]: You brought this up in slack, can this be quickly searched and we remove the TODO?

File: tap_workday_raas/discover.py (hunk 5 above)
Comment (asaf-erlich) [inline tap_workday_raas/discover.py]: Is this still untested? Can we add a test?

Comment (asaf-erlich) [reply tap_workday_raas/discover.py]: It looks like it is tested, so maybe just remove the TODO?

File: tap_workday_raas/client.py (hunk 1 above)
Comment (cosimon) [reply tap_workday_raas/client.py]: Done.

File: tap_workday_raas/client.py (hunk 2 above)
Comment (cosimon) [reply tap_workday_raas/client.py]: Reworked into a streaming parser.

File: tap_workday_raas/discover.py (hunk 3 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Done.

File: tap_workday_raas/discover.py (hunk 4 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Done, but the transformer treats it as a number, ignoring the format.

File: tap_workday_raas/discover.py (hunk 5 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: This is to-done with tests in `discovery_tests.py`.

File: tap_workday_raas/discover.py (hunk 6 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Done.

File: tap_workday_raas/client.py (hunk 7 above)
Comment (cosimon) [reply tap_workday_raas/client.py:11]: If the split key is missing then the string is returned in a length 1 array, it cannot have a length < 1.

```
>>> x = 'foo'
>>> x.split('?')
['foo']
```

File: tap_workday_raas/client.py (hunk 2 above)
Comment (cosimon) [reply tap_workday_raas/client.py]: Now iteratively parses the stream.

File: tap_workday_raas/discover.py (hunk 5 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Created a test that covers the function in `discovery_tests.py`.

File: tap_workday_raas/discover.py [hunk 8]
```diff
@@ -11,8 +11,12 @@
 def _element_to_schema(element):
     elem_type = element.attrib['type'].split(':')[1]
     is_nullable = element.attrib.get('minOccurs') == '0'
-    # NB: The only actual value I've seen for maxOccurs is "unbounded". There may be other possible values
-    is_list = element.attrib.get("maxOccurs") == 'unbounded'
+
+    max_occurs = elem.attrib.get("maxOccurs")
+    if max_occurs not in (None, "unbounded"):
+        raise Exception("Found unexpected value for maxOccurs attribute.")
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: This should probably include the value as well.

File: tap_workday_raas/discover.py [hunk 9]
```diff
@@ -25,7 +29,7 @@ def _element_to_schema(element):
         #TODO What else do we need for decimal types?
         schema = {
             'type': ['number'],
-            'format': 'decimal'
+            'format': 'singer-decimal'
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: For singer-decimal, the type must be `string`, and we should confirm that the Transformer allows it.

Given that folks are using it, I think changing this should be reserved for a major version bump. Having them properly typed as `Decimal`s will cause inconsistent types.

TLDR - The `format` isn't a thing for `number` JSON types, so it should just be removed.

File: tap_workday_raas/discover.py [hunk 10]
```diff
@@ -81,14 +84,17 @@ def generate_schema_for_report(xsd):
         # When elem's type attribute is a type defined as its own complexType - a nested object
         if elem_type in complex_type_mapping:
 
-            # TODO Check if the element is a list and type it correctly
-            is_list = elem.attrib.get("maxOccurs") == 'unbounded'
+            max_occurs = elem.attrib.get("maxOccurs")
+            if max_occurs not in (None, "unbounded"):
+                raise Exception("Found unexpected value for maxOccurs attribute.")
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: Again, it should probably put the value in the exception here for traceability.

**Review by dmosorast — CHANGES_REQUESTED** [review]
Almost there 😬

File: tap_workday_raas/discover.py (hunk 10 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Good call. Included it.

File: tap_workday_raas/discover.py (hunk 9 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Removed and changed to just `'type': ['number']`

File: tap_workday_raas/discover.py [hunk 11]
```diff
@@ -86,7 +86,7 @@ def generate_schema_for_report(xsd):
 
             max_occurs = elem.attrib.get("maxOccurs")
             if max_occurs not in (None, "unbounded"):
-                raise Exception("Found unexpected value for maxOccurs attribute.")
+                raise Exception("Found unexpected value for maxOccurs attribute: '{}'".format(maxOccurs))
```
Comment (dmosorast) [inline tap_workday_raas/discover.py]: Should be `max_occurs`

Comment (dmosorast) [reply tap_workday_raas/discover.py]: And there should be another spot with this I think?

**Review by dmosorast — CHANGES_REQUESTED** [review]

File: tap_workday_raas/discover.py (hunk 11 above)
Comment (cosimon) [reply tap_workday_raas/discover.py]: Done.

**Review by dmosorast — APPROVED** [review]
